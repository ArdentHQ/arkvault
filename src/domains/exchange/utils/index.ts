export * from "./assertions";
export * from "./errors";
