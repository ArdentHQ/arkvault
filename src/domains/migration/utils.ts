import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
