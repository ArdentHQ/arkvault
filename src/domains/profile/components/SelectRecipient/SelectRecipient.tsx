import { Networks } from "@ardenthq/sdk";
import { Contracts } from "@ardenthq/sdk-profiles";
import cn from "classnames";
import React, { useCallback, useEffect, useMemo, useRef, useState } from "react";

import { useTranslation } from "react-i18next";
import { Address } from "@/app/components/Address";
import { Avatar } from "@/app/components/Avatar";
import { Circle } from "@/app/components/Circle";
import { useFormField } from "@/app/components/Form/useFormField";
import { Icon } from "@/app/components/Icon";
import { Select } from "@/app/components/SelectDropdown";
import { TruncateEnd } from "@/app/components/TruncateEnd";
import { useWalletAlias, WalletAliasResult } from "@/app/hooks/use-wallet-alias";
import { AddressProperties, useProfileAddresses } from "@/domains/profile/hooks/use-profile-addresses";
import { SearchRecipient } from "@/domains/transaction/components/SearchRecipient";

type SelectRecipientProperties = {
	network?: Networks.Network;
	address?: string;
	profile: Contracts.IProfile;
	disabled?: boolean;
	isInvalid?: boolean;
	showOptions?: boolean;
	showWalletAvatar?: boolean;
	contactSearchTitle?: string;
	contactSearchDescription?: string;
	placeholder?: string;
	exceptMultiSignature?: boolean;
	onChange?: (address: string | undefined, alias: WalletAliasResult) => void;
} & Omit<React.InputHTMLAttributes<any>, "onChange">;

const ProfileAvatar = ({ address }: any) => {
	if (!address) {
		return (
			<Circle
				className="border-theme-secondary-200 bg-theme-secondary-200 dark:border-theme-secondary-700 dark:bg-theme-secondary-700"
				size="sm"
				noShadow
			/>
		);
	}
	return <Avatar address={address} size="sm" noShadow />;
};

const OptionLabel = ({
	option,
	network,
	profile,
}: {
	option: any;
	network?: Networks.Network;
	profile: Contracts.IProfile;
}) => {
	const address = option.value;

	const { getWalletAlias } = useWalletAlias();

	const { alias } = useMemo(
		() =>
			getWalletAlias({
				address,
				network,
				profile,
			}),
		[address, getWalletAlias, network, profile],
	);

	return (
		<Address
			address={address}
			walletName={alias}
			addressClass={cn({ "text-theme-primary-600": !alias && option.isSelected })}
			walletNameClass={cn({ "text-theme-primary-600": option.isSelected })}
			truncateOnTable
		/>
	);
};

export const SelectRecipient = React.forwardRef<HTMLInputElement, SelectRecipientProperties>(
	(
		{
			address,
			profile,
			disabled,
			isInvalid,
			showOptions = true,
			showWalletAvatar = true,
			network,
			placeholder,
			exceptMultiSignature,
			onChange,
			contactSearchTitle,
			contactSearchDescription,
		}: SelectRecipientProperties,
		reference,
	) => {
		const { t } = useTranslation();

		const { getWalletAlias } = useWalletAlias();

		const [isRecipientSearchOpen, setIsRecipientSearchOpen] = useState(false);

		const selectRecipientReference = useRef<HTMLDivElement | null>(null);

		/*
		 * Initial value for selected address is set in useEffect below via onChangeAddress.
		 * This is made to also retrieve alias information of the selected address.
		 */
		const [selectedAddress, setSelectedAddress] = useState<string | undefined>();
		const [selectedAddressAlias, setSelectedAddressAlias] = useState<WalletAliasResult | undefined>();

		const fieldContext = useFormField();

		const onChangeAddress = useCallback(
			(addressValue: string | undefined, emitOnChange = true) => {
				if (addressValue === selectedAddress) {
					return;
				}

				const alias = getWalletAlias({
					address: addressValue ?? "",
					network,
					profile,
				});

				setSelectedAddress(addressValue);
				setSelectedAddressAlias(alias);

				if (emitOnChange) {
					onChange?.(addressValue, alias);
				}
			},
			[getWalletAlias, network, onChange, profile, selectedAddress],
		);

		const isInvalidValue = isInvalid || fieldContext?.isInvalid;

		// Modify the address from parent component
		useEffect(() => {
			onChangeAddress(address, false);
		}, [address]); // eslint-disable-line react-hooks/exhaustive-deps

		const { allAddresses } = useProfileAddresses({ network, profile }, exceptMultiSignature);

		const recipientOptions = allAddresses.map(({ address }: AddressProperties) => ({
			label: address,
			value: address,
		}));

		const openRecipients = useCallback(() => {
			if (disabled) {
				return;
			}

			const dropdownWrapper = selectRecipientReference.current!.querySelector(
				"[role=combobox]",
			) as HTMLDivElement;
			// Necessary to ensure the select dropdown is hidden
			if (dropdownWrapper.getAttribute("aria-expanded") === "true") {
				const input = selectRecipientReference.current!.querySelector(
					"input#SelectRecipient__dropdown-input",
				) as HTMLInputElement;
				input.focus();
				input.blur();
			}

			setIsRecipientSearchOpen(true);
		}, [disabled, selectRecipientReference]);

		const onAction = useCallback(
			(changedAddress: string) => {
				onChangeAddress(changedAddress);
				setIsRecipientSearchOpen(false);
			},
			[onChangeAddress, setIsRecipientSearchOpen],
		);

		return (
			<div>
				<div
					ref={selectRecipientReference}
					data-testid="SelectRecipient__wrapper"
					className="relative flex w-full items-center text-left"
				>
					<Select
						id="SelectRecipient__dropdown"
						showCaret={false}
						isInvalid={isInvalidValue}
						disabled={disabled}
						defaultValue={selectedAddress}
						placeholder={placeholder || t("COMMON.ADDRESS")}
						ref={reference}
						options={showOptions ? recipientOptions : []}
						showOptions={showOptions}
						allowFreeInput={true}
						onChange={(option: any) => onChangeAddress(option.value)}
						addons={{
							end: showOptions
								? {
										content: (
											<div
												data-testid="SelectRecipient__select-recipient"
												className={cn("flex items-center", { "cursor-pointer": !disabled })}
												onClick={openRecipients}
											>
												<Icon name="User" size="lg" />
											</div>
										),
									}
								: undefined,
							start: showWalletAvatar
								? {
										content: (
											<div className="flex items-center">
												<ProfileAvatar address={selectedAddress} />
												{!!selectedAddressAlias?.alias && (
													<TruncateEnd
														className="ml-2 font-semibold"
														text={selectedAddressAlias.alias}
														showTooltip
													/>
												)}
											</div>
										),
									}
								: undefined,
						}}
						renderLabel={(option) => <OptionLabel option={option} network={network} profile={profile} />}
					/>
				</div>

				<SearchRecipient
					title={contactSearchTitle}
					description={contactSearchDescription}
					isOpen={isRecipientSearchOpen}
					recipients={allAddresses}
					onAction={onAction}
					onClose={() => setIsRecipientSearchOpen(false)}
					selectedAddress={selectedAddress}
					profile={profile}
				/>
			</div>
		);
	},
);

SelectRecipient.displayName = "SelectRecipient";
