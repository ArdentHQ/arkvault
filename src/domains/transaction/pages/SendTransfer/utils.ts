export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
