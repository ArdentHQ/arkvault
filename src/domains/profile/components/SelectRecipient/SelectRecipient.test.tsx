/* eslint-disable @typescript-eslint/require-await */
import { Contracts } from "@ardenthq/sdk-profiles";
import userEvent from "@testing-library/user-event";
import React from "react";

import { env, getDefaultProfileId, render, screen, waitFor } from "@/utils/testing-library";

import { SelectRecipient } from "./SelectRecipient";

let profile: Contracts.IProfile;

const selectRecipient = () => screen.getByTestId("SelectRecipient__select-recipient");

describe("SelectRecipient", () => {
	beforeAll(() => {
		profile = env.profiles().findById(getDefaultProfileId());
	});

	it("should render empty", () => {
		const { container } = render(<SelectRecipient profile={profile} />);

		expect(container).toMatchSnapshot();
	});

	it("should render disabled", () => {
		const { container } = render(<SelectRecipient profile={profile} disabled />);

		expect(container).toMatchSnapshot();
	});

	it("should render invalid", () => {
		const { container } = render(<SelectRecipient profile={profile} isInvalid />);

		expect(container).toMatchSnapshot();
	});

	it("should update internal state when prop changes", () => {
		const { container, rerender } = render(<SelectRecipient profile={profile} />);

		rerender(<SelectRecipient profile={profile} address="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT" />);

		expect(container).toMatchSnapshot();
	});

	it("should open and close contacts modal", async () => {
		render(<SelectRecipient profile={profile} address="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT" />);

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();

		await userEvent.click(selectRecipient());

		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		await userEvent.click(screen.getByTestId("Modal__close-button"));

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();
	});

	it("should focus & blur the address input when is expanded", async () => {
		const contactsSpy = vi.spyOn(profile.contacts(), "findByAddress").mockReturnValue([]);

		render(<SelectRecipient profile={profile} />);
		const recipientInputField = screen.getByTestId("SelectDropdown__input");

		const focusSpy = vi.spyOn(recipientInputField, "focus");
		const blurSpy = vi.spyOn(recipientInputField, "blur");

		// 1. Focus the select input which opens the drodpown
		await userEvent.click(recipientInputField);

		// 2. Click to open the modal
		await userEvent.click(selectRecipient());

		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		expect(focusSpy).toHaveBeenCalledTimes(2);
		expect(blurSpy).toHaveBeenCalledTimes(2);

		contactsSpy.mockRestore();
		focusSpy.mockRestore();
		blurSpy.mockRestore();
	});

	it("should not focus & blur the address input when is not expanded", async () => {
		const contactsSpy = vi.spyOn(profile.contacts(), "findByAddress").mockReturnValue([]);

		render(<SelectRecipient profile={profile} />);
		const recipientInputField = screen.getByTestId("SelectDropdown__input");

		const focusSpy = vi.spyOn(recipientInputField, "focus");
		const blurSpy = vi.spyOn(recipientInputField, "blur");

		await userEvent.click(selectRecipient());

		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		expect(focusSpy).not.toHaveBeenCalled();
		expect(blurSpy).not.toHaveBeenCalled();

		contactsSpy.mockRestore();
		focusSpy.mockRestore();
		blurSpy.mockRestore();
	});

	it("should select address from contacts modal", async () => {
		render(<SelectRecipient profile={profile} address="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT" />);

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();

		await userEvent.click(selectRecipient());

		await expect(screen.findByTestId("Modal__inner")).resolves.toBeVisible();

		const firstContactAddress = screen.getByTestId("RecipientListItem__select-button-2");

		await userEvent.click(firstContactAddress);

		await waitFor(() => expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument());

		await waitFor(() =>
			expect(screen.getByTestId("SelectDropdown__input")).toHaveValue(
				profile.contacts().values()[0].addresses().values()[0].address(),
			),
		);
	});

	it("should not show select recipient button if showOptions is false", async () => {
		render(<SelectRecipient profile={profile} showOptions={false} />);

		expect(screen.queryByTestId("queryByTestId-recipient")).not.toBeInTheDocument();
	});

	it("should not open contacts modal if disabled", async () => {
		render(<SelectRecipient profile={profile} address="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT" disabled />);

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();

		await userEvent.click(selectRecipient());

		await waitFor(() => expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument());
	});

	it("should call onChange prop when entered address in input", async () => {
		const address = "bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT";

		const contactsSpy = vi.spyOn(profile.contacts(), "findByAddress").mockReturnValue([]);
		const onChange = vi.fn();

		render(<SelectRecipient profile={profile} onChange={onChange} />);
		const recipientInputField = screen.getByTestId("SelectDropdown__input");

		await userEvent.paste(recipientInputField, address);

		expect(screen.getByTestId("SelectDropdown__input")).toHaveValue(address);

		expect(onChange).toHaveBeenCalledWith(address, {
			alias: undefined,
			isContact: false,
			isDelegate: false,
		});

		contactsSpy.mockRestore();
	});

	it("should call onChange prop if provided", async () => {
		const onChange = vi.fn();

		render(
			<SelectRecipient
				profile={profile}
				onChange={onChange}
				address="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT"
			/>,
		);

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();

		await userEvent.click(selectRecipient());

		await waitFor(() => {
			expect(screen.getByTestId("Modal__inner")).toBeInTheDocument();
		});

		const firstAddress = screen.getByTestId("RecipientListItem__select-button-2");

		await userEvent.click(firstAddress);

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();

		const selectedAddressValue = profile.contacts().values()[0].addresses().values()[0].address();

		expect(screen.getByTestId("SelectDropdown__input")).toHaveValue(selectedAddressValue);
		expect(onChange).toHaveBeenCalledWith(selectedAddressValue, expect.any(Object));
	});

	it("should call onChange prop only when values change", async () => {
		const onChange = vi.fn();

		render(<SelectRecipient profile={profile} onChange={onChange} address="D61mfSggzbvQgTUe6JhYKH2doHaqJ3Dyib" />);

		const selectedAddressValue = profile.contacts().values()[0].addresses().values()[0].address();

		expect(screen.getByTestId("SelectDropdown__input")).toHaveValue(selectedAddressValue);

		await userEvent.click(selectRecipient());

		await waitFor(() => {
			expect(screen.getByTestId("Modal__inner")).toBeInTheDocument();
		});

		const lastAddress = screen.getByTestId("RecipientListItem__selected-button-2");

		await userEvent.click(lastAddress);

		expect(screen.getByTestId("SelectDropdown__input")).toHaveValue(selectedAddressValue);
		expect(onChange).not.toHaveBeenCalled();
	});

	it("should filter recipients list by network if provided", async () => {
		const function_ = vi.fn();

		const [wallet] = profile.wallets().findByCoinWithNetwork("ARK", "ark.devnet");

		render(
			<SelectRecipient
				profile={profile}
				onChange={function_}
				address="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT"
				network={wallet.coin().network()}
			/>,
		);

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();

		await userEvent.click(selectRecipient());

		await waitFor(() => expect(screen.queryByTestId("RecipientListItem__select-button")).not.toBeInTheDocument());
	});

	it("should filter recipients list by MultiSignature type", async () => {
		const { rerender } = render(<SelectRecipient profile={profile} />);

		expect(screen.queryByTestId("Modal__inner")).not.toBeInTheDocument();

		await userEvent.click(selectRecipient());

		await waitFor(() => {
			expect(screen.getByTestId("Modal__inner")).toBeInTheDocument();
		});

		expect(screen.getAllByTestId("TableRow")).toHaveLength(6);

		const isMultiSignatureSpy = vi
			.spyOn(profile.wallets().first(), "isMultiSignature")
			.mockImplementation(() => true);

		rerender(<SelectRecipient profile={profile} exceptMultiSignature />);

		await userEvent.click(selectRecipient());

		await waitFor(() => {
			expect(screen.getByTestId("Modal__inner")).toBeInTheDocument();
		});

		expect(screen.getAllByTestId("TableRow")).toHaveLength(1);

		isMultiSignatureSpy.mockRestore();
	});
});
