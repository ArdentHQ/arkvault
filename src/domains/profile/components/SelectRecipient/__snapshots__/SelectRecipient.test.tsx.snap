// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectRecipient > should render disabled 1`] = `
<div>
  <div>
    <div
      class="relative flex w-full items-center text-left"
      data-testid="SelectRecipient__wrapper"
    >
      <div
        class="relative w-full"
        data-testid="SelectDropdown"
      >
        <div
          class="sr-only css-14zfvme"
        >
          <div
            class="relative flex h-full flex-1"
          >
            <input
              autocomplete="off"
              class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base css-1937ogs"
              data-testid="select-list__input"
              readonly=""
              tabindex="-1"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="SelectRecipient__dropdown-menu"
            role="combobox"
          >
            <div>
              <label
                for="SelectRecipient__dropdown-input"
                id="SelectRecipient__dropdown-label"
              />
              <div
                class="css-4lfin5"
                disabled=""
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="border-theme-secondary-200 bg-theme-secondary-200 dark:border-theme-secondary-700 dark:bg-theme-secondary-700 ring-theme-background duration-100 css-cm9d5u"
                  />
                </div>
                <div
                  class="relative flex h-full flex-1"
                >
                  <input
                    aria-autocomplete="list"
                    aria-controls="SelectRecipient__dropdown-menu"
                    aria-labelledby="SelectRecipient__dropdown-label"
                    autocomplete="off"
                    class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base cursor-default text-theme-secondary-text css-1937ogs"
                    data-testid="SelectDropdown__input"
                    disabled=""
                    id="SelectRecipient__dropdown-input"
                    placeholder="Address"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-primary-300 dark:text-theme-secondary-600"
                  data-testid="Input__addon-end"
                >
                  <div
                    class=""
                  >
                    <div
                      class="flex items-center"
                      data-testid="SelectRecipient__select-recipient"
                    >
                      <div
                        class="css-1i4b0eq"
                        height="20"
                        width="20"
                      >
                        <svg
                          id="user"
                          viewBox="0 0 20 20"
                          x="0"
                          xml:space="preserve"
                          xmlns="http://www.w3.org/2000/svg"
                          y="0"
                        >
                          <g
                            fill="none"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          >
                            <circle
                              cx="10"
                              cy="5"
                              r="4"
                            />
                            <path
                              d="M2 19c0-4.899 3.544-9 8-9 4.354 0 8 3.987 8 9"
                            />
                          </g>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="SelectRecipient__dropdown-label"
              id="SelectRecipient__dropdown-menu"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SelectRecipient > should render empty 1`] = `
<div>
  <div>
    <div
      class="relative flex w-full items-center text-left"
      data-testid="SelectRecipient__wrapper"
    >
      <div
        class="relative w-full"
        data-testid="SelectDropdown"
      >
        <div
          class="sr-only css-14zfvme"
        >
          <div
            class="relative flex h-full flex-1"
          >
            <input
              autocomplete="off"
              class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base css-1937ogs"
              data-testid="select-list__input"
              readonly=""
              tabindex="-1"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="SelectRecipient__dropdown-menu"
            role="combobox"
          >
            <div>
              <label
                for="SelectRecipient__dropdown-input"
                id="SelectRecipient__dropdown-label"
              />
              <div
                class="css-14zfvme"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="border-theme-secondary-200 bg-theme-secondary-200 dark:border-theme-secondary-700 dark:bg-theme-secondary-700 ring-theme-background duration-100 css-cm9d5u"
                  />
                </div>
                <div
                  class="relative flex h-full flex-1"
                >
                  <input
                    aria-autocomplete="list"
                    aria-controls="SelectRecipient__dropdown-menu"
                    aria-labelledby="SelectRecipient__dropdown-label"
                    autocomplete="off"
                    class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base cursor-default css-1937ogs"
                    data-testid="SelectDropdown__input"
                    id="SelectRecipient__dropdown-input"
                    placeholder="Address"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-primary-300 dark:text-theme-secondary-600"
                  data-testid="Input__addon-end"
                >
                  <div
                    class=""
                  >
                    <div
                      class="flex items-center cursor-pointer rounded bg-transparent p-1 text-theme-secondary-700 transition-colors hover:bg-theme-primary-100 hover:text-theme-primary-700 dark:text-theme-secondary-600 dark:hover:bg-theme-secondary-700 dark:hover:text-white"
                      data-testid="SelectRecipient__select-recipient"
                    >
                      <div
                        class="css-1i4b0eq"
                        height="20"
                        width="20"
                      >
                        <svg
                          id="user"
                          viewBox="0 0 20 20"
                          x="0"
                          xml:space="preserve"
                          xmlns="http://www.w3.org/2000/svg"
                          y="0"
                        >
                          <g
                            fill="none"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          >
                            <circle
                              cx="10"
                              cy="5"
                              r="4"
                            />
                            <path
                              d="M2 19c0-4.899 3.544-9 8-9 4.354 0 8 3.987 8 9"
                            />
                          </g>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="SelectRecipient__dropdown-label"
              id="SelectRecipient__dropdown-menu"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SelectRecipient > should render invalid 1`] = `
<div>
  <div>
    <div
      class="relative flex w-full items-center text-left"
      data-testid="SelectRecipient__wrapper"
    >
      <div
        class="relative w-full"
        data-testid="SelectDropdown"
      >
        <div
          class="sr-only css-1njetft"
        >
          <div
            class="relative flex h-full flex-1"
          >
            <input
              aria-invalid="true"
              autocomplete="off"
              class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base css-1937ogs"
              data-testid="select-list__input"
              readonly=""
              tabindex="-1"
              type="text"
              value=""
            />
          </div>
          <div
            class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-danger-500"
            data-testid="Input__addon-end"
          >
            <span
              data-testid="Input__error"
            >
              <div
                class="text-theme-danger-500 css-1i4b0eq"
                height="20"
                width="20"
              >
                <svg
                  id="circle-exclamation-mark"
                  viewBox="0 0 20 20"
                  x="0"
                  xml:space="preserve"
                  xmlns="http://www.w3.org/2000/svg"
                  y="0"
                >
                  <g
                    fill="none"
                    stroke="currentColor"
                    stroke-width="2"
                  >
                    <circle
                      cx="10"
                      cy="10"
                      r="9"
                      stroke-miterlimit="10"
                    />
                    <path
                      d="M10 10.5v-5"
                      stroke-linecap="round"
                      stroke-miterlimit="10"
                    />
                    <path
                      d="M10 14h0"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                    />
                  </g>
                </svg>
              </div>
            </span>
          </div>
        </div>
        <div
          class="w-full"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="SelectRecipient__dropdown-menu"
            role="combobox"
          >
            <div>
              <label
                for="SelectRecipient__dropdown-input"
                id="SelectRecipient__dropdown-label"
              />
              <div
                class="css-1njetft"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="border-theme-secondary-200 bg-theme-secondary-200 dark:border-theme-secondary-700 dark:bg-theme-secondary-700 ring-theme-background duration-100 css-cm9d5u"
                  />
                </div>
                <div
                  class="relative flex h-full flex-1"
                >
                  <input
                    aria-autocomplete="list"
                    aria-controls="SelectRecipient__dropdown-menu"
                    aria-invalid="true"
                    aria-labelledby="SelectRecipient__dropdown-label"
                    autocomplete="off"
                    class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base cursor-default css-1937ogs"
                    data-testid="SelectDropdown__input"
                    id="SelectRecipient__dropdown-input"
                    placeholder="Address"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-danger-500"
                  data-testid="Input__addon-end"
                >
                  <span
                    data-testid="Input__error"
                  >
                    <div
                      class="text-theme-danger-500 css-1i4b0eq"
                      height="20"
                      width="20"
                    >
                      <svg
                        id="circle-exclamation-mark"
                        viewBox="0 0 20 20"
                        x="0"
                        xml:space="preserve"
                        xmlns="http://www.w3.org/2000/svg"
                        y="0"
                      >
                        <g
                          fill="none"
                          stroke="currentColor"
                          stroke-width="2"
                        >
                          <circle
                            cx="10"
                            cy="10"
                            r="9"
                            stroke-miterlimit="10"
                          />
                          <path
                            d="M10 10.5v-5"
                            stroke-linecap="round"
                            stroke-miterlimit="10"
                          />
                          <path
                            d="M10 14h0"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                          />
                        </g>
                      </svg>
                    </div>
                  </span>
                  <div
                    class="pl-3"
                  >
                    <div
                      class="flex items-center cursor-pointer rounded bg-transparent p-1 text-theme-secondary-700 transition-colors hover:bg-theme-primary-100 hover:text-theme-primary-700 dark:text-theme-secondary-600 dark:hover:bg-theme-secondary-700 dark:hover:text-white"
                      data-testid="SelectRecipient__select-recipient"
                    >
                      <div
                        class="css-1i4b0eq"
                        height="20"
                        width="20"
                      >
                        <svg
                          id="user"
                          viewBox="0 0 20 20"
                          x="0"
                          xml:space="preserve"
                          xmlns="http://www.w3.org/2000/svg"
                          y="0"
                        >
                          <g
                            fill="none"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          >
                            <circle
                              cx="10"
                              cy="5"
                              r="4"
                            />
                            <path
                              d="M2 19c0-4.899 3.544-9 8-9 4.354 0 8 3.987 8 9"
                            />
                          </g>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="SelectRecipient__dropdown-label"
              id="SelectRecipient__dropdown-menu"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`SelectRecipient > should update internal state when prop changes 1`] = `
<div>
  <div>
    <div
      class="relative flex w-full items-center text-left"
      data-testid="SelectRecipient__wrapper"
    >
      <div
        class="relative w-full"
        data-testid="SelectDropdown"
      >
        <div
          class="sr-only css-14zfvme"
        >
          <div
            class="relative flex h-full flex-1"
          >
            <input
              autocomplete="off"
              class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base css-1937ogs"
              data-testid="select-list__input"
              readonly=""
              tabindex="-1"
              type="text"
              value="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT"
            />
          </div>
        </div>
        <div
          class="w-full"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="SelectRecipient__dropdown-menu"
            role="combobox"
          >
            <div>
              <label
                for="SelectRecipient__dropdown-input"
                id="SelectRecipient__dropdown-label"
              />
              <div
                class="css-14zfvme"
              >
                <div
                  class="flex items-center"
                >
                  <div
                    class="relative inline-flex items-center justify-center rounded-full align-middle transition-all duration-100 h-8 w-8 text-sm shrink-0"
                    data-testid="Avatar"
                  >
                    <div
                      class="inline-flex h-full w-full items-center justify-center overflow-hidden align-middle rounded-full"
                    >
                      <img
                        alt="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT"
                        src="data:image/svg+xml;utf8,<svg version="1.1" xmlns="http://www.w3.org/2000/svg" class="picasso" width="100" height="100" viewBox="0 0 100 100"><style>.picasso circle{mix-blend-mode:soft-light;}</style><rect fill="rgb(255, 193, 7)" width="100" height="100"/><circle r="45" cx="20" cy="40" fill="rgb(244, 67, 54)"/><circle r="35" cx="100" cy="70" fill="rgb(233, 30, 99)"/><circle r="55" cx="70" cy="50" fill="rgb(139, 195, 74)"/></svg>"
                        title="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT"
                      />
                    </div>
                  </div>
                </div>
                <div
                  class="relative flex h-full flex-1"
                >
                  <input
                    aria-autocomplete="list"
                    aria-controls="SelectRecipient__dropdown-menu"
                    aria-labelledby="SelectRecipient__dropdown-label"
                    autocomplete="off"
                    class="no-ligatures w-full border-none !text-sm placeholder:text-theme-secondary-400 dark:placeholder:text-theme-secondary-700 sm:!text-base css-1937ogs"
                    data-testid="SelectDropdown__input"
                    id="SelectRecipient__dropdown-input"
                    placeholder="Address"
                    type="text"
                    value="bP6T9GQ3kqP6T9GQ3kqP6T9GQ3kqTTTP6T9GQ3kqT"
                  />
                </div>
                <div
                  class="flex items-center space-x-3 divide-x divide-theme-secondary-300 dark:divide-theme-secondary-800 text-theme-primary-300 dark:text-theme-secondary-600"
                  data-testid="Input__addon-end"
                >
                  <div
                    class=""
                  >
                    <div
                      class="flex items-center cursor-pointer rounded bg-transparent p-1 text-theme-secondary-700 transition-colors hover:bg-theme-primary-100 hover:text-theme-primary-700 dark:text-theme-secondary-600 dark:hover:bg-theme-secondary-700 dark:hover:text-white"
                      data-testid="SelectRecipient__select-recipient"
                    >
                      <div
                        class="css-1i4b0eq"
                        height="20"
                        width="20"
                      >
                        <svg
                          id="user"
                          viewBox="0 0 20 20"
                          x="0"
                          xml:space="preserve"
                          xmlns="http://www.w3.org/2000/svg"
                          y="0"
                        >
                          <g
                            fill="none"
                            stroke="currentColor"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          >
                            <circle
                              cx="10"
                              cy="5"
                              r="4"
                            />
                            <path
                              d="M2 19c0-4.899 3.544-9 8-9 4.354 0 8 3.987 8 9"
                            />
                          </g>
                        </svg>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              aria-labelledby="SelectRecipient__dropdown-label"
              id="SelectRecipient__dropdown-menu"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
